### DISCUSSION TITLE
Training the SpanCategorizer via the Update method

### DISCUSSION THREAD

**Discussion by Romulus-5** [explosion/spaCy#10480] [Help: Coding & Implementations] [answered]
Currently, we've been experimenting with the excellent new SpanCategorizer pipeline in preparation for starting a more extensive annotation work with Prodigy. 

We are currently using the Update method. However, although the training happens for all iterations, the resulting model does not make any predictions. I'm guessing this is an issue with formatting the Examples training instances. In particular, we are unclear on storing the annotated spans in the Example instance.

    nlp = spacy.blank('en')
    spancat = nlp.add_pipe("spancat")

    for l in labels:
        spancat.add_label(l)
    nlp.begin_training()

    for itn in range(50): #50 iterations
        losses = {}

        for raw_text, annotations in train_data:
            doc = nlp.make_doc(raw_text)
            spans = list()

            for a in annotations:
                #Annotations is a list of lists: [starting character index, ending character index, label].
                span = doc.char_span(a[0], a[1], label=a[2], alignment_mode="expand")
                spans.append(span)

            doc.spans["sc"] = spans
            example = Example.from_dict(doc, {})
            nlp.update([example], drop=0.2, losses=losses)

**Comment by polm** [comment, marked as answer]
Without getting into exactly what is going wrong here, is there a reason you're not using config-based training? We strongly recommend not writing your own training loop in order to avoid errors. If you don't want to use the config, we also expose the internal training loop as a function you can use.

Also, it looks like something went wrong when you pasted your code and the indentation is off and maybe some parts are missing - you don't use `entity_offsets` and it's not clear where `annotations` is defined. Can you double-check your code?

**Reply by Romulus-5** [reply]
Thanks for the quick reply and sorry for the copy-paste confusion. I have fixed the code. 

We started experimenting with our training loop because it seemed easier to test: we thought we were getting the test annotations wrong. However, by trying with the configuration, we could make it work. Thank you very much.
